sw/device/tests/entropy_src_fw_override_test.c

```diff
@@ -0,0 +1,407 @@
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "entropy_src_regs.h"                         // Generated.
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"  // Generated.
+
+OTTF_DEFINE_TEST_CONFIG();
+
+enum {
+  kEntropySrcHealthTestWindowSize = 0x200,
+  /**
+   * Observe FIFO threshold: half of the FIFO size.
+   */
+  kEntropySrcFifoThreshold = 32,
+  /**
+   * The number of contiguous samples we want to capture.
+   */
+  kContiguousSamplesCount = 1024,
+  /**
+   * Number of bits per sample.
+   */
+  kBitsPerSample = 4,
+  /**
+   * Timeout for operations which wait for random data.
+   */
+  kTimeoutUsec = 50000,
+  /**
+   * Size of buffer in words to hold all the samples, assuming
+   * 4-bit samples at the most.
+   */
+  kFifoBufferSizeWords =
+      kContiguousSamplesCount * kBitsPerSample / sizeof(uint32_t),
+};
+
+static dif_aes_t aes;
+static dif_csrng_t csrng;
+static dif_edn_t edn0;
+static dif_edn_t edn1;
+static dif_entropy_src_t entropy_src;
+static dif_hmac_t hmac;
+static dif_kmac_t kmac;
+static dif_rv_core_ibex_t rv_core_ibex;
+static dif_rv_plic_t rv_plic;
+
+// Buffers to collect samples into.
+//
+// We use a double-buffering approach, where one buffer is for the "input"
+// samples that we read from the observe FIFO, and the other is for "output"
+// samples that we need to write to the conditioner.
+//
+// When the input buffer is filled and the output buffer is empty, we swap them
+// and the bytes we collected are written out to the conditioner. The input
+// buffer is now empty, and we can start writing new samples to it.
+static uint32_t sample_buf_0[kFifoBufferSizeWords];
+static uint32_t sample_buf_1[kFifoBufferSizeWords];
+
+// These pointers track which buffer is the input or the output and are cursors
+// into where we have written to / read from them so far.
+static uint32_t *input_buf = sample_buf_0;
+static uint32_t *output_buf = sample_buf_1 + kFifoBufferSizeWords;
+
+// These counters track how many words are remaining to be written to the input
+// buffer or read out of the output buffer.
+static size_t words_to_input = kFifoBufferSizeWords;
+static size_t words_to_output = 0;
+
+static dif_edn_auto_params_t edn_params0;
+static dif_edn_auto_params_t edn_params1;
+
+/**
+ * Determine whether the observe FIFO has overflowed.
+ *
+ * TODO(#21279) Normally, one would rely on the FW_OV_RD_FIFO_OVERFLOW
+ * register but due to an RTL bug, the overflow bit is pulsed
+ * instead of latched so we cannot rely on it. Instead, rely
+ * on OBSERVE_FIFO_DEPTH and assume that if the FIFO is full
+ * then it has overflowed.
+ */
+bool entropy_src_fifo_has_overflowed(void) {
+  uint32_t fifo_depth;
+  CHECK_DIF_OK(dif_entropy_src_get_fifo_depth(&entropy_src, &fifo_depth));
+  return fifo_depth == ENTROPY_SRC_PARAM_OBSERVE_FIFO_DEPTH;
+}
+
+/**
+ * Drain observe FIFO.
+ */
+static void drain_observe_fifo(void) {
```

Comment:
maybe not necessary for this PR but we could consider moving to the testutils now that it is used in two tests